[WIP] Zinc + Persistent Bazel Worker Processes

ahirreddy opened the pull request:
This is a WIP PR that integrates Zinc with Bazel's support for persistent [worker processes](http://bazel.io/blog/2015/12/10/java-workers.html). The main advantage of running with persistent Zinc workers is taking advantage of warm Scala compilers across builds (this includes reusing warm compiler instances across different build targets). I've found significant compile time improvements on large builds (a 3x speed up for our internal build), although I've done no concrete bench marking.

This PR does not enable the incremental compiler. It is unclear if it makes sense to enable incremental re-compilation within Bazel, as I do not believe the results of incremental compilation are guaranteed to be reproducible.

To use the persistent worker use the `scala_worker` rule and run:
`bazel build --strategy=Scala=worker --worker_max_instances=<n> <target>`.

You can also add these default parameters to your `.bazelrc` (which can be global or per repo) so that you don't have to constantly type them:
```
build --strategy=Scala=worker --worker_max_instances=8
```

Notes:
  - It uses a modified version of Zinc at the moment, which can read arguments from a file instead of the command line. This was originally to deal with issues where we exceeded the maximum size argument list that could be passed as a command line parameter. It's unclear if this is still necessary since Bazel communicates with the worker process using protobuf over STDIN. The custom version can be found [here](https://github.com/ahirreddy/zinc/tree/file-arguments).
  - I had to use the `java_binary` rule instead of the `scala_binary` because the scala rule changes the present working directory before invoking the JVM. The issue here is that Bazel provides input paths and arguments to the worker process as relative paths that only resolve in the initial working directory.
  - To test the changes against our internal repo, I had to add 2.10 support. Minus one hack that I will clean up, it is straight forward to support 2.10 and 2.11 with the same rules.
  - I pulled the Worker implementation from the ExampleWorker in Bazel, and converted it to support Scala.

ahirreddy commented on scala/scala.bzl:
```diff
@@ -100,18 +86,123 @@ touch -t 198001010000 {manifest}
+    progress_message="scala ijar %s" % ctx.label,)
+
+
+def _compile(ctx, jars, buildijar, usezinc):
```
Refactored this a bit so that it can be reused for both scalac and zinc invocations

johnynek commented on scala/scala.bzl:
```diff
@@ -254,7 +348,8 @@ _implicit_deps = {
   "_scalalib": attr.label(default=Label("@scala//:lib/scala-library.jar"), single_file=True, allow_files=True),
-  "_scalaxml": attr.label(default=Label("@scala//:lib/scala-xml_2.11-1.0.4.jar"), single_file=True, allow_files=True),
+  # "_scalaxml": attr.label(default=Label("@scala//:lib/scala-xml_2.11-1.0.4.jar"), single_file=True, allow_files=True),
+  "_scalaxml": attr.label(default=Label("@scala//:lib/scala-library.jar"), single_file=True, allow_files=True),
```
what is this change about?

johnynek replied:
ahh, was this a hack to work around 2.11 being hardwired in here?

ahirreddy replied:
Exactly. I need to find a better way to handle this.

johnynek commented on scala/scala.bzl:
```diff
@@ -369,3 +550,68 @@ def scala_repositories():
+def zinc_repositories():
+  native.http_file(
+    name = "zinc",
+    url = "https://databricks-mvn.s3.amazonaws.com/binaries/zinc/12-29-15/zinc?AWSAccessKeyId=AKIAJ6V3VSHTA5RSYEQA&Expires=1482970631&Signature=tAP2QKWEnte6v3DpRFtAxGNkUps%3D",
```
where are these jars coming from? Why can't we use a standard location?

ahirreddy replied:
These were links being used for our internal build. This is the customized version of zinc that I mentioned above. I don't believe that we actually need it anymore, so I can just pull zinc from typesafe or github. I'll just have to update the PR to be able to work with standard zinc instead of my fork

johnynek commented on scala/scala.bzl:
```diff
@@ -369,3 +550,68 @@ def scala_repositories():
+  )
+
+  native.http_jar(
+    name = "scala_compiler_jar",
```
why can't we use the scala_compiler in the tgz above?

johnynek commented on scala/scala.bzl:
```diff
@@ -369,3 +550,68 @@ def scala_repositories():
+  )
+
+  native.http_jar(
+    name = "scala_library_jar",
```
why can't we use the scala_library in the tgz above?

johnynek commented on scala/scala.bzl:
```diff
@@ -369,3 +550,68 @@ def scala_repositories():
+  )
+
+  native.http_jar(
+    name = "scala_reflect_jar",
```
same comment about scala_reflect. Can't we get it from the tgz?

damienmg commented:
Note that the author of this PR hasn't signed the CLA. I will meet with our open-source office on Friday and ask them if we need Google CLA for this repo. Until then we cannot accept this PR unless the author  sign the CLA.

johnynek commented on scala/scala.bzl:
```diff
@@ -273,6 +368,67 @@ _common_attrs = {
+      "main_class": attr.string(),
+      "exports": attr.label_list(allow_files=False),
+      # Worker Args
+      "worker": attr.label(
```
is this a private label? Should users ever change this? If private, can we use `_worker`?

johnynek commented on scala/scala.bzl:
```diff
@@ -369,3 +550,68 @@ def scala_repositories():
     sha256 = "f198967436a5e7a69cfd182902adcfbcb9f2e41b349e1a5c8881a2407f615962",
   )
+
+def scala_2_10_repositories():
```
this is tricky. We probably want to think about how to support different versions of scala. I assume this zinc is only working for 2.10 now?

johnynek commented on scala/scala.bzl:
```diff
@@ -273,6 +368,67 @@ _common_attrs = {
+
+}
+
+scala_worker = rule(
```
can we not make this an option for the other rules? Seems like using `scala_worker` requires more knowledge than we would want everyone to have to have to use it.

johnynek commented on scala/ScalaWorker.scala:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    // Options copied from Nailgun.scala in Zinc
+    val options = List("-cp", classpath, "-server", "-Xms1024m", "-Xmx3g", "-XX:MaxPermSize=384m",
```
can these be passed in somehow?

johnynek commented on scala/ScalaWorker.scala:
```diff
@@ -0,0 +1,220 @@
+            )
+          } catch {
+            case e: Exception =>
+              // We use System.out.println as not to accidentally write to real stdout
```
I don't follow this comment. How is it not real stdout?

Can we not use stderr here?

johnynek replied:
ahh,, I see, you change it above.

johnynek commented on scala/scala.bzl:
```diff
@@ -273,6 +368,67 @@ _common_attrs = {
+      default=Label("@zinc_0_3_10_SNAPSHOT_jar//jar"),
+      single_file=True,
+      allow_files=True),
+  "_scala_compiler_jar": attr.label(
```
can we not share the code from the scala .tar.gz and point there instead?

johnynek commented:
related to #8

johnynek commented:
Thanks so much for working on this! Very exciting!

Open issues:
1. how do we cleanly support scala 2.10 & 2.11 (and soon 2.12) without copying too much code?
2. is zinc reliable (and faster) in this mode than just using a warm scalac? Could we just use scalac as a worker by calling the main method on that? I think in bazel we definitely don't want to break the invariant that identical code produces identical jars. How can we test that here? Can we make a stand alone test for the scala worker?
3. Can we have just one mode? Ideally, we have a faster compile, and that is just the default. Maybe users can opt out in some memory constained cases with an option? Could we make such an option a WORKSPACE setting?
4. Can we get the tests green on travis?

damienmg commented:
Btw, /cc @philwo FYI, a PR to use worker for Scala/Zinc :)

johnynek commented on scala/ScalaWorker.scala:
```diff
@@ -0,0 +1,220 @@
+
+        for (input <- request.getInputsList().asScala) {
+          inputs.put(input.getPath(), input.getDigest().toStringUtf8())
+        }
```
nit: idiomatic scala would be:
```scala
request.getInputsList.asScala.foreach { input =>
  inputs.put(input.getPath, input.getDigest.toStringUtf8)
}
```

johnynek commented on scala/ScalaWorker.scala:
```diff
@@ -0,0 +1,220 @@
+        }
+
+        val baos = new ByteArrayOutputStream()
+        var exitCode = 0
```
can we have `val exitCode = try { ...` below rather than a var? var adds cognitive load to reviewing, which IMO should only be used when necessary for perf. In this case `val` is just as good.

johnynek commented on scala/ScalaWorker.scala:
```diff
@@ -0,0 +1,220 @@
+
+          try {
+            clientArgs = request.getArgumentsList.asScala.flatMap { arg =>
+              // srcjars must be extracted before we can pass them to zinc
```
these aren't supported yet by the other rules.

johnynek commented on scala/ScalaWorker.scala:
```diff
@@ -0,0 +1,220 @@
+              }
+            }
+            awaitServer()
+            exitCode = zincClient.run(
```
this could just be the return value rather than changing the var.

johnynek commented on scala/ScalaWorker.scala:
```diff
@@ -0,0 +1,220 @@
+              }
+
+              e.printStackTrace()
+              exitCode = 1
```
could just return this rather than using a var.

johnynek commented:
@damienwo @philwo are there docs about such things as how many worker processes are started? I assume more than one. Maybe as many as there are CPUs?

johnynek commented:
related to #14

ahirreddy commented:
@johnynek Thanks for taking a look! I'll start addressing the comments in the next few days when I have some free time.

@damienmg Will sign the CLA once I get approval on my end.

johnynek commented:
@ahirreddy great!

I'd be interested to see if we can just jump in here:
https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/Main.scala

and make an invocation. For instance:
```scala
@annotation.tailrec
def loop(): Unit = {
  val args = readArgsFromProto
  val comp = new MainClass
  // see: https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/Driver.scala#L37
  comp.process(args)
  if (comp.reporter.hasErrors) reportErrors(comp.reporter) else reportSuccess()
  loop()
}
```
might do it, and as long as there are no static vars, that should be exactly like running scalac except we get the jit to warm up.

What do you think of that?

damienmg commented:
@ahirreddy Thanks

I have meet with our Open-source office and they confirmed that we need CLA for everything that is contributed to a bazelbuild repository

philwo commented:
@johnynek Currently, Bazel starts up to 4 workers per "key", where a key is a unique combination of command-line to start the worker, environment variables, working directory and Spawn mnemonic. (In other words, if you use both the Javac persistent worker and the Scala persistent workers, there will be up to 4 workers spawned for each.)

You can tune that via the --worker_max_instances flag: https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/worker/WorkerOptions.java#L48

philwo commented:
@ahirreddy I'd love to see this in Bazel - is there anything we could do to help you get it ready for merging?

ahirreddy commented:
Hey @philwo I'll try to get to it this weekend!

ulfjack commented:
Any updates on this?

xynny commented:
we are using this internally right now.  I work with ahir.  There is currently an issue with the caching as well because the zinc worker emits a temporary directory instead of a jar.  This breaks caching because it can't generate a digest for a directory as an input.  I'm working on this issue right now.

johnynek commented:
I'm pessimistic about the incremental compiler aspect of `zinc`. The pants folks have had a lot of correctness bugs around it. I think that should definitely be its own separate plugin for people who like to live on the edge.

That said, I think doing a worker process just using the normal compiler should be pretty easy and we can reinstantiate a new compiler on each call, and I think the JIT will give us a pretty good win.

We are using bazel in the mode where CI does not to a full rebuild, so we care a lot about the correctness part and would not want to trade that for a slightly faster compile.

ittaiz commented:
+1 on correctness.
We've had zinc issues to caching which caused some people to lose faith.
On שבת, 14 במאי 2016 at 3:25 P. Oscar Boykin <notifications@github.com>
wrote:

> I'm pessimistic about the incremental compiler aspect of zinc. The pants
> folks have had a lot of correctness bugs around it. I think that should
> definitely be its own separate plugin for people who like to live on the
> edge.
>
> That said, I think doing a worker process just using the normal compiler
> should be pretty easy and we can reinstantiate a new compiler on each call,
> and I think the JIT will give us a pretty good win.
>
> We are using bazel in the mode where CI does not to a full rebuild, so we
> care a lot about the correctness part and would not want to trade that for
> a slightly faster compile.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/12#issuecomment-219188578>
>

philwo commented:
+1 on favoring correctness over speed and leaving out the incremental compilation. JIT and no more JVM start-up overhead should already give a pretty nice performance boost in itself.

ittaiz commented:
@xynny @ahirreddy 
Any chance of splitting the work to persistent worker with incremental compilation off and then another one in a plugin like @johnynek suggested?

ahirreddy commented:
This PR doesn't actually use the incremental compiler component of Zinc, so we have no issues around correctness or reproducibility. Zinc just happens to provide a nice persistent build server wrapper around the Scala compiler. Non reproducible builds on our end have usually been caused by timestamp issues in jars.

I've actually worked on this alot internally, adding code coverage support and support for Scala 2.10 and 2.11 (another Zinc benefit is that it can host Scala 2.10 and 2.11 compilers simultaneously). I'll try to push these upstream once I have some time.

bazel-io commented:
Can one of the admins verify this patch?

ittaiz commented:
Hi @ahirreddy,
Any news?

philwo commented:
I'd also be interested in how this project is going :) It would be an awesome and very welcome contribution to Bazel.

johnynek commented:
closed by #91
